### PR TITLE
T11: Close Thread when switching Folder

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #1582

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 1]
```diff
@@ -75,6 +75,11 @@ open class TwoPaneFragment : Fragment() {
                 R.string.noConversationSelected,
                 displayedFolder.getLocalizedName(requireContext()),
             )
+
+            if (displayedFolder.id != previousFolderId) {
+                previousFolderId = displayedFolder.id
+                if (isInThread) closeThread()
+            }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:82]: We should probably poll to know if we want to do this or not

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 2]
```diff
@@ -51,31 +52,43 @@ open class TwoPaneFragment : Fragment() {
...
 
-            val displayedFolder = (if (isInSearch) {
-                Folder().apply { name = getString(R.string.searchFolderName) }
+            val currentFolder = if (isInSearch) {
+                searchFolder
             } else {
-                folder
-            }) ?: return@observe
+                folder ?: return@observe
+            }
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:81]: That logic can be in the `ViewModel`. That way, you only need to return the folder.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 3]
```diff
@@ -51,31 +52,43 @@ open class TwoPaneFragment : Fragment() {
...
             rightPaneFolderName.value = getString(
                 R.string.noConversationSelected,
-                displayedFolder.getLocalizedName(requireContext()),
+                currentFolder.getLocalizedName(requireContext()),
             )
+
+            if (currentFolder.id != previousFolderId) {
+                previousFolderId = currentFolder.id
+                if (isInThread) closeThread()
+            }
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:91]: You can have this logic in a separate LiveData in order to close the thread.

**Review by tevincent — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 4]
```diff
@@ -201,11 +201,13 @@ class MainViewModel @Inject constructor(
 
     //region Current Thread
     val currentThreadUid = MutableLiveData<String?>(null)
-
     val closeThreadTrigger = SingleLiveEvent<Unit>()
 
+    val isInThread get() = currentThreadUid.value != null
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:206]: You can inline this

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 1 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt]: Done. We keep closing it.

**Review by FabianDevel — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 5]
```diff
@@ -50,28 +51,40 @@ abstract class TwoPaneFragment : Fragment() {
...
 
-            val displayedFolder = (if (this@TwoPaneFragment is SearchFragment) {
-                Folder().apply { name = getString(R.string.searchFolderName) }
+            val currentFolder = if (this@TwoPaneFragment is SearchFragment) {
+                searchFolder
             } else {
-                folder
-            }) ?: return@observe
+                folder ?: return@observe
+            }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:80]: Can't we inline this?

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 6]
```diff
@@ -50,28 +51,40 @@ abstract class TwoPaneFragment : Fragment() {
     fun isOnlyLeftShown() = slidingPaneLayout.let { it.isSlideable && !it.isOpen }
     fun isOnlyRightShown() = slidingPaneLayout.let { it.isSlideable && it.isOpen }
 
+    private val searchFolder by lazy {
+        Folder().apply {
+            id = FolderController.SEARCH_FOLDER_ID
+            name = getString(R.string.searchFolderName)
+        }
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:59]: Is it not strange to create a search Folder manually like this? Wouldn't a Pair or a data class be less confusing?

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 6 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:59]: We need a Folder object in several places for the rest of the logic.

Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt]: Fixed.

**Review by LunarX — APPROVED** [review]

**Pull request merged**